File: packages/shared/src/channels.ts
// ============================================================================
// Channel Constants
// ============================================================================

export const CHANNEL_PREFIX = {
  PTY_INPUT: 'pty.input',
  PTY_OUTPUT: 'pty.output',
  PTY_CONTROL: 'pty.control',
  TN3270_INPUT: 'tn3270.input',
  TN3270_OUTPUT: 'tn3270.output',
  TN3270_CONTROL: 'tn3270.control',
  SESSIONS: 'sessions.events',
  GATEWAY_CONTROL: 'gateway.control',
} as const;

export type ChannelPrefix = (typeof CHANNEL_PREFIX)[keyof typeof CHANNEL_PREFIX];

export function getPtyInputChannel(sessionId: string): string {
  return `${CHANNEL_PREFIX.PTY_INPUT}.${sessionId}`;
}

export function getPtyOutputChannel(sessionId: string): string {
  return `${CHANNEL_PREFIX.PTY_OUTPUT}.${sessionId}`;
}

export function getPtyControlChannel(sessionId: string): string {
  return `${CHANNEL_PREFIX.PTY_CONTROL}.${sessionId}`;
}

export function getTn3270InputChannel(sessionId: string): string {
  return `${CHANNEL_PREFIX.TN3270_INPUT}.${sessionId}`;
}

export function getTn3270OutputChannel(sessionId: string): string {
  return `${CHANNEL_PREFIX.TN3270_OUTPUT}.${sessionId}`;
}

export function getTn3270ControlChannel(): string {
  return CHANNEL_PREFIX.TN3270_CONTROL;
}

export function getGatewayControlChannel(): string {
  return CHANNEL_PREFIX.GATEWAY_CONTROL;
}

export function getSessionsChannel(): string {
  return CHANNEL_PREFIX.SESSIONS;
}

export function parseChannel(channel: string): {
  prefix: string;
  sessionId: string | null;
} {
  const parts = channel.split('.');
  if (parts.length >= 3) {
    return {
      prefix: `${parts[0]}.${parts[1]}`,
      sessionId: parts.slice(2).join('.'),
    };
  }
  return {
    prefix: channel,
    sessionId: null,
  };
}

File: gateway/src/core/channels.py
# ============================================================================
# Valkey Channel Helpers (matching @terminal/shared)
# ============================================================================

"""
Channel naming conventions:
- gateway.control        - Global control channel for PTY session creation
- tn3270.control         - Global control channel for TN3270 session creation
- pty.input.<id>         - Input to PTY (keystrokes from user)
- pty.output.<id>        - Output from PTY (terminal output)
- pty.control.<id>       - Control messages (resize, destroy)
- tn3270.input.<id>      - Input to TN3270 (keystrokes from user)
- tn3270.output.<id>     - Output from TN3270 (terminal output)
"""


def get_pty_input_channel(session_id: str) -> str:
    """Get the input channel for a PTY session."""
    return f"pty.input.{session_id}"


def get_pty_output_channel(session_id: str) -> str:
    """Get the output channel for a PTY session."""
    return f"pty.output.{session_id}"


def get_pty_control_channel(session_id: str) -> str:
    """Get the control channel for a PTY session."""
    return f"pty.control.{session_id}"


def get_tn3270_input_channel(session_id: str) -> str:
    """Get the input channel for a TN3270 session."""
    return f"tn3270.input.{session_id}"


def get_tn3270_output_channel(session_id: str) -> str:
    """Get the output channel for a TN3270 session."""
    return f"tn3270.output.{session_id}"


# Control channel for all PTY gateway instances
GATEWAY_CONTROL_CHANNEL = "gateway.control"

# Control channel for TN3270 gateway instances
TN3270_CONTROL_CHANNEL = "tn3270.control"
